### ISSUE TITLE
Uninitialized value $ExtUtils::ParseXS::num in subtraction

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12085]
Migrated from [rt.perl.org#112776](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112776) (status was 'resolved')

Searchable as RT112776$

**Comment by p5pRT** [comment]
### From perl@cjmweb.net

This is a bug report for perl from perl@&#8203;cjmweb\.net\,
generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.




---
I'm working on a Perl module using XS to bind to a C library\. When I
build it\, I'm getting a warning message saying

&nbsp;  Use of uninitialized value $num in subtraction \(\-\) at
&nbsp;  /usr/lib/perl5/vendor\_perl/5\.14\.2/ExtUtils/ParseXS\.pm line 1769\,
&nbsp;  \<GEN8> line 90\.

This is triggered by the code generated by ExtUtils&#8203;::Constant\.
Commenting out the INCLUDE&#8203;: const\-xs\.inc line in Foo\.xs removes the
warning\. I don't know whether the bug is in ExtUtils&#8203;::ParseXS or
in ExtUtils&#8203;::Constant\, but even if EU&#8203;::C is generating bogus code\,
EU&#8203;::PXS shouldn't be emitting warnings like that\.

I'm using ExtUtils&#8203;::Constant 0\.23\, ExtUtils&#8203;::ParseXS 3\.15\,
Module&#8203;::Build 0\.38\, and Perl 5\.14\.2\.

I've managed to pare it down to a reasonably small test case that
doesn't require any external C library\.  I've placed it in a GitHub
repo&#8203;: https://github.com/madsen/ParseXS-ExtUtils-Constant-bug
and attached a copy\.

To reproduce the bug\, clone the repo\, type "perl Build\.PL" and then
"\./Build"\. You should see&#8203;:

&nbsp;  $ perl Build\.PL
&nbsp;  Regenerating constants\.\.\.
&nbsp;  Created MYMETA\.yml and MYMETA\.json
&nbsp;  Creating new 'Build' script for 'Foo' version '0\.01'
&nbsp;  $ \./Build
&nbsp;  Building Foo
&nbsp;  Use of uninitialized value $num in subtraction \.\.\.
&nbsp;  gcc \.\.\. \-o lib/Foo\.o lib/Foo\.c
&nbsp;  ExtUtils&#8203;::Mkbootstrap&#8203;::Mkbootstrap\('blib/arch/auto/Foo/Foo\.bs'\)
&nbsp;  gcc \.\.\. \-o blib/arch/auto/Foo/Foo\.so lib/Foo\.o

You'll only see "Regenerating constants\.\.\." if you have
ExtUtils&#8203;::Constant installed\. You shouldn't need it to reproduce the
bug\, because I've added the generated files to the repo\.

Whatever the problem is\, it doesn't seem to stop the code from
working\, because the included test does pass\.

I've asked about this on http&#8203;://stackoverflow\.com/q/10378986/8355
but nobody's come up with an answer\.



---


---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=low
&nbsp;    module=ExtUtils&#8203;::ParseXS


---
Site configuration information for perl 5\.14\.2&#8203;:

Configured by Gentoo at Sat Oct 22 18&#8203;:25&#8203;:36 CDT 2011\.

Summary of my perl5 \(revision 5 version 14 subversion 2\) configuration&#8203;:

&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.37\-gentoo\-r4\, archname=i686\-linux
&nbsp;    uname='linux byte 2\.6\.37\-gentoo\-r4 \#2 smp preempt tue may 10
18&#8203;:56&#8203;:04 cdt 2011 i686 intel\(r\) core\(tm\)2 duo cpu e6750 @&#8203; 2\.66ghz
genuineintel gnulinux '
&nbsp;    config\_args='\-des \-Duseshrplib \-Darchname=i686\-linux
\-Dcc=i686\-pc\-linux\-gnu\-gcc \-Doptimize=\-march=core2 \-O2 \-pipe \-ggdb
\-Dldflags=\-Wl\,\-O1 \-Wl\,\-\-as\-needed \-Dprefix=/usr \-Dsiteprefix=/usr
\-Dvendorprefix=/usr \-Dscriptdir=/usr/bin \-Dprivlib=/usr/lib/perl5/5\.14\.2
\-Darchlib=/usr/lib/perl5/5\.14\.2/i686\-linux
\-Dsitelib=/usr/local/lib/perl5/5\.14\.2
\-Dsitearch=/usr/local/lib/perl5/5\.14\.2/i686\-linux
\-Dvendorlib=/usr/lib/perl5/vendor\_perl/5\.14\.2
\-Dvendorarch=/usr/lib/perl5/vendor\_perl/5\.14\.2/i686\-linux
\-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3
\-Dsiteman1dir=/usr/share/man/man1 \-Dsiteman3dir=/usr/share/man/man3
\-Dvendorman1dir=/usr/share/man/man1 \-Dvendorman3dir=/usr/share/man/man3
\-Dman1ext=1 \-Dman3ext=3pm \-Dlibperl=libperl\.so\.5\.14\.2 \-Dlocincpth=
\-Duselargefiles \-Dd\_semctl\_semun \-Dcf\_by=Gentoo \-Dmyhostname=localhost
\-Dperladmin=root@&#8203;localhost \-Dinstallusrbinperl=n \-Ud\_csh \-Uusenm
\-Di\_ndbm \-Di\_gdbm \-Di\_db \-DDEBUGGING=\-g
\-Dinc\_version\_list=5\.14\.0/i686\-linux 5\.14\.0 5\.14\.1/i686\-linux 5\.14\.1 '
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='i686\-pc\-linux\-gnu\-gcc'\, ccflags ='\-fno\-strict\-aliasing \-pipe
\-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-march=core2 \-O2 \-pipe \-ggdb'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-pipe'
&nbsp;    ccversion=''\, gccversion='4\.5\.3'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='i686\-pc\-linux\-gnu\-gcc'\, ldflags ='\-Wl\,\-O1 \-Wl\,\-\-as\-needed
\-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=/lib/libc\-2\.11\.3\.so\, so=so\, useshrplib=true\,
libperl=libperl\.so\.5\.14\.2
&nbsp;    gnulibc\_version='2\.11\.3'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-march=core2 \-O2 \-pipe \-ggdb
\-L/usr/local/lib \-Wl\,\-O1 \-Wl\,\-\-as\-needed'

Locally applied patches&#8203;:
&nbsp;    0001\-gentoo\_MakeMaker\-RUNPATH\.diff
&nbsp;    0002\-gentoo\_MakeMaker\-delete\_packlist\.diff
&nbsp;    0003\-gentoo\_config\_over\.diff
&nbsp;    0004\-gentoo\_cpan\_definstalldirs\.diff
&nbsp;    0005\-gentoo\_cpanplus\_definstalldirs\.diff
&nbsp;    0006\-gentoo\_create\-libperl\-soname\.diff
&nbsp;    0007\-gentoo\_drop\-fstack\-protector\.diff
&nbsp;    0008\-gentoo\_enc2xs\.diff
&nbsp;    0009\-gentoo\_mod\-paths\.diff



---
@&#8203;INC for perl 5\.14\.2&#8203;:
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl5/5\.14\.2/i686\-linux
&nbsp;    /usr/local/lib/perl5/5\.14\.2
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.14\.2/i686\-linux
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.14\.2
&nbsp;    /usr/local/lib/perl5
&nbsp;    /usr/lib/perl5/vendor\_perl
&nbsp;    /usr/lib/perl5/5\.14\.2/i686\-linux
&nbsp;    /usr/lib/perl5/5\.14\.2
&nbsp;    \.



---
Environment for perl 5\.14\.2&#8203;:
&nbsp;    HOME=/home/cjm
&nbsp;    LANG=en\_US\.utf8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)

PATH=/home/cjm/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/opt/bin&#8203;:/usr/i686\-pc\-linux\-gnu/gcc\-bin/4\.5\.3&#8203;:/usr/games/bin&#8203;:/var/qmail/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_UNICODE=SAL
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From perl@cjmweb.net

[testcase.tar.gz](https://rt-archive.perl.org/perl5/Ticket/Attachment/1113478/560446/testcase.tar.gz)

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat May 05 12&#8203;:37&#8203;:25 2012\, cjm wrote&#8203;:
> This is a bug report for perl from perl@&#8203;cjmweb\.net\,
> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> I'm working on a Perl module using XS to bind to a C library\. When I
> build it\, I'm getting a warning message saying
> 
>   Use of uninitialized value $num in subtraction \(\-\) at
>   /usr/lib/perl5/vendor\_perl/5\.14\.2/ExtUtils/ParseXS\.pm line 1769\,
>   \<GEN8> line 90\.
> 
> This is triggered by the code generated by ExtUtils&#8203;::Constant\.
> Commenting out the INCLUDE&#8203;: const\-xs\.inc line in Foo\.xs removes the
> warning\. I don't know whether the bug is in ExtUtils&#8203;::ParseXS or
> in ExtUtils&#8203;::Constant\, but even if EU&#8203;::C is generating bogus code\,
> EU&#8203;::PXS shouldn't be emitting warnings like that\.
> 
> I'm using ExtUtils&#8203;::Constant 0\.23\, ExtUtils&#8203;::ParseXS 3\.15\,
> Module&#8203;::Build 0\.38\, and Perl 5\.14\.2\.
> 

Do you get this warning if you use the \(older\) version of
ExtUtils&#8203;::ParseXS that came with Perl 5\.14\.2 \-\- v2\.2210?

jimk

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat May 05 15&#8203;:15&#8203;:09 2012\, jkeenan wrote&#8203;:
> 
> Do you get this warning if you use the \(older\) version of
> ExtUtils&#8203;::ParseXS that came with Perl 5\.14\.2 \-\- v2\.2210?
> 

I answered my own question&#8203;:  Yes\.

\#\#\#\#\#
$ \./Build
Building Foo
Use of uninitialized value $ExtUtils&#8203;::ParseXS&#8203;::num in subtraction \(\-\) at
/usr/local/lib/perl5/5\.14\.2/ExtUtils/ParseXS\.pm line 1780\, \<GEN8> line 90\.
\#\#\#\#\#

In the version you are using\, 3\.15\, modify line 1117 of
ExtUtils/ParseXS\.pm from this&#8203;:

&nbsp;   1117     $self\->\{var\_num\} = $self\->\{args\_match\}\->\{$var\_name\};

to&#8203;:

&nbsp;            $self\->\{var\_num\} = \($self\->\{args\_match\}\->\{$var\_name\}\) || 0;

That should eliminate the uninitialized value warning\.  However\, I
suspect we should proceed farther back than that point to see why we
assumed that the RHS of line 1117 would always be initialized\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From perl@cjmweb.net

On Sat May 05 15&#8203;:38&#8203;:10 2012\, jkeenan wrote&#8203;:
> In the version you are using\, 3\.15\, modify line 1117 of
> ExtUtils/ParseXS\.pm

That won't do much good for me\, since the people installing my module
won't have the patch\, and they're the ones who are most likely to be
concerned about it\.  I can just ignore the warning\.

I guess I'll just have to include a note about it in the README\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat May 05 16&#8203;:05&#8203;:47 2012\, cjm wrote&#8203;:
> On Sat May 05 15&#8203;:38&#8203;:10 2012\, jkeenan wrote&#8203;:
> > In the version you are using\, 3\.15\, modify line 1117 of
> > ExtUtils/ParseXS\.pm
> 
> That won't do much good for me\, since the people installing my module
> won't have the patch\, and they're the ones who are most likely to be
> concerned about it\.  I can just ignore the warning\.
> 
> I guess I'll just have to include a note about it in the README\.

I was only asking you to do that for our diagnostic purposes\.  Your
\(presumably\) legitimate use\-case has uncovered a bug in ExtUtils&#8203;::ParseXS\.

If I import Data&#8203;::Dumper into ParseXS\.pm and modify ExtUtils&#8203;::ParseXS
3\.15 at the point described above like this&#8203;:

\#\#\#\#\#
print STDERR Dumper $self\->\{args\_match\};
print STDERR "\\$var\_name&#8203;: $var\_name\\n";
&nbsp;    $self\->\{var\_num\} = $self\->\{args\_match\}\->\{$var\_name\} || 0;
\#\#\#\#\#

\.\.\. then try to build your module\, I get this output&#8203;:

\#\#\#\#\#
Regenerating constants\.\.\.
Created MYMETA\.yml and MYMETA\.json
Creating new 'Build' script for 'Foo' version '0\.01'
Building Foo
$VAR1 = \{
&nbsp;  'sv' => 1
\};
$var\_name&#8203;: sv
$VAR1 = \{
&nbsp;  'sv' => 1
\};
$var\_name&#8203;: s
cc \-I/usr/local/lib/perl5/5\.14\.2/darwin\-2level/CORE \-DXS\_VERSION="0\.01"
\-DVERSION="0\.01" \-c \-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe
\-I/usr/local/include \-I/opt/local/include \-O3 \-o lib/Foo\.o lib/Foo\.c
ExtUtils&#8203;::Mkbootstrap&#8203;::Mkbootstrap\('blib/arch/auto/Foo/Foo\.bs'\)
env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc \-bundle \-undefined dynamic\_lookup
\-L/usr/local/lib \-L/opt/local/lib \-o blib/arch/auto/Foo/Foo\.bundle lib/Foo\.o
\#\#\#\#\#

So the second time through a loop\, there is a difference between the key
of the sole element in $self\->\{args\_match\} and the content of $var\_name\.
 This is what can leave $self\->\{args\_match\}\->\{$var\_name\} undefined\.

However\, I don't know/remember enough of ParseXS to say what that means\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From perl@cjmweb.net

On Sat May 05 17&#8203;:56&#8203;:48 2012\, jkeenan wrote&#8203;:
> I was only asking you to do that for our diagnostic purposes\.  Your
> \(presumably\) legitimate use\-case has uncovered a bug in ExtUtils&#8203;::ParseXS\.

Sorry\, I misunderstood\.  Yes\, adding "|| 0" to that line removes the
warning\.

**Comment by p5pRT** [comment]
### From @jkeenan

Please review the patch attached\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>0001-Diagnosis-of-uninitialized-value-warning-reported-by.patch</summary>

```diff
From f8c3bbd4e2315ec1599f7dc346dd1c00836479bf Mon Sep 17 00:00:00 2001
From: jkeenan <jkeenan@cpan.org>
Date: Sat, 5 May 2012 23:31:52 -0400
Subject: [PATCH] Diagnosis of uninitialized value warning reported by Christopher J. Madsen++.

The set of keys found in the hash $self->{args_found} fundamentally derives
from the parsing of $func_header inside process_file().  The members of that
set are fixed by the time INPUT_handler() is called for the first time within
that subroutine.  They therefore do not necessarily include every instance of
$var_name; that variable only exists within INPUT_handler().
$self->{args_match}->{$var_name} does not necessarily exist.  So, in order to
be able to use $self->{var_num} in arithmetic expressions, we have to provide
it with a defined default value of 0.

For RT #112776.
---
 dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm b/dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm
index 883d905..19c46c0 100644
--- a/dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm
+++ b/dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm
@@ -1114,7 +1114,7 @@ sub INPUT_handler {
       print "\t" . map_type($self, $var_type, undef);
       $printed_name = 0;
     }
-    $self->{var_num} = $self->{args_match}->{$var_name};
+    $self->{var_num} = $self->{args_match}->{$var_name} || 0;
 
     if ($self->{var_num}) {
       my $typemap = $self->{typemap}->get_typemap(ctype => $var_type);
-- 
1.6.3.2


```
</details>

**Comment by p5pRT** [comment]
### From perl@cjmweb.net

On Sat May 05 20&#8203;:53&#8203;:41 2012\, jkeenan wrote&#8203;:
> Please review the patch attached\.

Well\, I can confirm that it solves my problem\, but I don't know enough
about ExtUtils&#8203;::ParseXS to be sure it's the right solution\, or if it's
just masking the real problem\.

**Comment by p5pRT** [comment]
### From @tsee

On 05/06/2012 05&#8203;:58 AM\, Christopher J\. Madsen via RT wrote&#8203;:
> On Sat May 05 20&#8203;:53&#8203;:41 2012\, jkeenan wrote&#8203;:
>> Please review the patch attached\.
>
> Well\, I can confirm that it solves my problem\, but I don't know enough
> about ExtUtils&#8203;::ParseXS to be sure it's the right solution\, or if it's
> just masking the real problem\.

That is precisely why I hadn't included a simple change like the 
proposed patch yet \-\- I am also not entirely certain about whether it's 
a fix or just plaster\. Jim's finding of key \(var name\) mismatches does 
make it sound like there's a deeper issue\.

I'm on the fence on whether it's better to apply the patch to avoid 
confusing users or whether it's better not to in the hope that one of 
the affected users cares as much as Christopher and figures it out\. 
Alas\, I don't currently have the time to dig in myself\.

\-\-Steffen

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, May 08\, 2012 at 07&#8203;:47&#8203;:48AM \+0200\, Steffen Mueller wrote&#8203;:
> On 05/06/2012 05&#8203;:58 AM\, Christopher J\. Madsen via RT wrote&#8203;:
> > On Sat May 05 20&#8203;:53&#8203;:41 2012\, jkeenan wrote&#8203;:
> >> Please review the patch attached\.
> >
> > Well\, I can confirm that it solves my problem\, but I don't know enough
> > about ExtUtils&#8203;::ParseXS to be sure it's the right solution\, or if it's
> > just masking the real problem\.
> 
> That is precisely why I hadn't included a simple change like the 
> proposed patch yet \-\- I am also not entirely certain about whether it's 
> a fix or just plaster\. Jim's finding of key \(var name\) mismatches does 
> make it sound like there's a deeper issue\.
> 
> I'm on the fence on whether it's better to apply the patch to avoid 
> confusing users or whether it's better not to in the hope that one of 
> the affected users cares as much as Christopher and figures it out\. 
> Alas\, I don't currently have the time to dig in myself\.

I fear that it's plaster\, and hence applying it does nothing to solve the
actual problem\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue\, May 08\, 2012 at 10&#8203;:13&#8203;:58AM \+0100\, Nicholas Clark wrote&#8203;:
> On Tue\, May 08\, 2012 at 07&#8203;:47&#8203;:48AM \+0200\, Steffen Mueller wrote&#8203;:
> > On 05/06/2012 05&#8203;:58 AM\, Christopher J\. Madsen via RT wrote&#8203;:
> > > On Sat May 05 20&#8203;:53&#8203;:41 2012\, jkeenan wrote&#8203;:
> > >> Please review the patch attached\.
> > >
> > > Well\, I can confirm that it solves my problem\, but I don't know enough
> > > about ExtUtils&#8203;::ParseXS to be sure it's the right solution\, or if it's
> > > just masking the real problem\.
> > 
> > That is precisely why I hadn't included a simple change like the 
> > proposed patch yet \-\- I am also not entirely certain about whether it's 
> > a fix or just plaster\. Jim's finding of key \(var name\) mismatches does 
> > make it sound like there's a deeper issue\.
> > 
> > I'm on the fence on whether it's better to apply the patch to avoid 
> > confusing users or whether it's better not to in the hope that one of 
> > the affected users cares as much as Christopher and figures it out\. 
> > Alas\, I don't currently have the time to dig in myself\.
> 
> I fear that it's plaster\, and hence applying it does nothing to solve the
> actual problem\.

There's one or both of two problems here&#8203;:

a\) the generated Build forces $^W on\, which isn't the normal state
when the code is run within xsubpp

b\) the undefined value itself is caused because the&#8203;:

&nbsp;        const char \*    s = SvPV\(sv\, len\);

in const\-xs\.inc is for a name that isn't a parameter\.

I can produce the same warning with code like&#8203;:

\#include "EXTERN\.h"
\#include "perl\.h"
\#include "XSUB\.h"

\#include "ppport\.h"

MODULE = Foo  PACKAGE = Foo  PREFIX = LIBMTP\_

PROTOTYPES&#8203;: DISABLE

void
foo\(\)
&nbsp;    const char \*s = "world";
&nbsp;  CODE&#8203;:
&nbsp;    printf\("hello %s\\n"\, s\);

I suspect the better change would be to change&#8203;:

&nbsp;  my $arg = "ST\(" \. \($num \- 1\) \. "\)";

to something along the lines of&#8203;:

&nbsp;    my $arg = $num ? "ST\(" \. \($num \- 1\) \. "\)"
&nbsp;       : qq\(die "$var isn't an argument\\\\n"\);

Tony

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue May 08 06&#8203;:34&#8203;:12 2012\, tonyc wrote&#8203;:
> On Tue\, May 08\, 2012 at 10&#8203;:13&#8203;:58AM \+0100\, Nicholas Clark wrote&#8203;:
> > On Tue\, May 08\, 2012 at 07&#8203;:47&#8203;:48AM \+0200\, Steffen Mueller wrote&#8203;:
> > > On 05/06/2012 05&#8203;:58 AM\, Christopher J\. Madsen via RT wrote&#8203;:
> > > > On Sat May 05 20&#8203;:53&#8203;:41 2012\, jkeenan wrote&#8203;:
> > > >> Please review the patch attached\.
> > > >
> > > > Well\, I can confirm that it solves my problem\, but I don't know
enough
> > > > about ExtUtils&#8203;::ParseXS to be sure it's the right solution\, or
if it's
> > > > just masking the real problem\.
> > > 
> > > That is precisely why I hadn't included a simple change like the 
> > > proposed patch yet \-\- I am also not entirely certain about whether
it's 
> > > a fix or just plaster\. Jim's finding of key \(var name\) mismatches
does 
> > > make it sound like there's a deeper issue\.
> > > 
> > > I'm on the fence on whether it's better to apply the patch to avoid 
> > > confusing users or whether it's better not to in the hope that one of 
> > > the affected users cares as much as Christopher and figures it out\. 
> > > Alas\, I don't currently have the time to dig in myself\.
> > 
> > I fear that it's plaster\, and hence applying it does nothing to
solve the
> > actual problem\.
> 
> There's one or both of two problems here&#8203;:
> 
> a\) the generated Build forces $^W on\, which isn't the normal state
> when the code is run within xsubpp
> 
> b\) the undefined value itself is caused because the&#8203;:
> 
>         const char \*    s = SvPV\(sv\, len\);
> 
> in const\-xs\.inc is for a name that isn't a parameter\.
> 
> I can produce the same warning with code like&#8203;:
> 
> \#include "EXTERN\.h"
> \#include "perl\.h"
> \#include "XSUB\.h"
> 
> \#include "ppport\.h"
> 
> MODULE = Foo  PACKAGE = Foo  PREFIX = LIBMTP\_
> 
> PROTOTYPES&#8203;: DISABLE
> 
> void
> foo\(\)
>     const char \*s = "world";
>   CODE&#8203;:
>     printf\("hello %s\\n"\, s\);
> 
> I suspect the better change would be to change&#8203;:
> 
>   my $arg = "ST\(" \. \($num \- 1\) \. "\)";
> 
> to something along the lines of&#8203;:
> 
>     my $arg = $num ? "ST\(" \. \($num \- 1\) \. "\)"
>        : qq\(die "$var isn't an argument\\\\n"\);

Just a reminder\. :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue Jun 19 21&#8203;:58&#8203;:01 2012\, sprout wrote&#8203;:
> On Tue May 08 06&#8203;:34&#8203;:12 2012\, tonyc wrote&#8203;:
> > I suspect the better change would be to change&#8203;:
> > 
> >   my $arg = "ST\(" \. \($num \- 1\) \. "\)";
> > 
> > to something along the lines of&#8203;:
> > 
> >     my $arg = $num ? "ST\(" \. \($num \- 1\) \. "\)"
> >        : qq\(die "$var isn't an argument\\\\n"\);
> 
> Just a reminder\. :\-\)

Fixed in 3f305fad20ca6851a9cb5d25370ab5719e7f184e and
115e60bcab0a2cba27b137c61175493a347d0adc

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
